src/parser/matches/arg_matches.rs

```diff
@@ -1379,38 +1529,103 @@ impl Default for RawValues<'_> {
 
 #[derive(Clone)]
 #[allow(missing_debug_implementations)]
-pub struct GroupedValues<'a> {
+pub struct GroupedValuesRef<'a, T> {
+    iter: Map<Iter<'a, Vec<AnyValue>>, fn(&Vec<AnyValue>) -> Vec<&T>>,
+}
+
+impl<'a, T> Iterator for GroupedValuesRef<'a, T>
+where
+    Self: 'a,
+{
+    type Item = Vec<&'a T>; // should this be an iterator instead?
```

Comment:
Yes to the comment
- I'd like to avoid the `collect` into a `Vec`
- I'd rather have opaque types.  All of the new getters intentionally have wrapper types, even if a `&Vec` would have worked

Reply:
Ok, so there are a few ways we could do this:

1. Add new iterator types for the inner iterators for each of these grouped iterators. I'm not really sure what to call them, maybe something like `GroupValues`, `GroupValuesRef`, and `RawGroupValues`? It will also add quite a bit of somewhat repetitive code (maybe a macro could help reduce the boilerplate?)
2. Use the existing iterator types for the inner types (for example, this line would be changed to `Item = ValuesRef<'a, T>` . The problem is that those existing types wrap a `std::iter::Flatten` type, whereas, the grouped iterators don't flatten the iterator. There are in turn a couple things we could do about that, none of them super great.
    
    1.  Use dynamic dispatch with a `dyn` type for the iterato inside, `Values`, `ValuesRef`, and `RawValues`. For example, the iter value in `ValuesRef` would become either `Map<&'a dyn Iterator<Item=&'a AnyValue> + DoubleEndedIterator, fn(&AnyValue) -> &T` or just `&'a dyn Iterator<Item=&'a T> + DoubleEndedIterator`. `Values` would use a `Box<dyn ...>` instead of a `&'a dyn ...`. 
    2. Create a new helper enum that looks something like:
       ```rust
       enum MaybeFlatten<I: Iterator> {
           Flatten(Flaten<I>),
           NotFlatten(I),
       }
       ```
       and then implement `Iterator`, and `DoubleEndedIterator` on that.
    3. Use the existing types as is, but use another level of vec to get the flatten in ther, basically we would do something like `vec![values].iter().flatten()`  to get it in the right type. 
 5. Just use a `Box<dyn Iterator<&'a T> + DoubleEndedIterator + ExactSizeIterator>`

Any thoughts on what the best approach is?

Reply:
Yeah, let's just create new iterator types.

Reply:
Do you think it would be worth moving all these iterators into a separate module/file?

Reply:
Maybe, but not as part of this PR so its easier to review.